// uploadthing/client/index.cjs
Object.defineProperty(exports, '__esModule', { value: true });

var Arr = require('effect/Array');
var Cause = require('effect/Cause');
var Console = require('effect/Console');
var Effect = require('effect/Effect');
var Function = require('effect/Function');
var Option = require('effect/Option');
var Runtime = require('effect/Runtime');
var shared = require('@uploadthing/shared');
var stdEnv = require('std-env');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return n;
}

var Arr__namespace = /*#__PURE__*/_interopNamespace(Arr);
var Cause__namespace = /*#__PURE__*/_interopNamespace(Cause);
var Console__namespace = /*#__PURE__*/_interopNamespace(Console);
var Effect__namespace = /*#__PURE__*/_interopNamespace(Effect);
var Option__namespace = /*#__PURE__*/_interopNamespace(Option);
var Runtime__namespace = /*#__PURE__*/_interopNamespace(Runtime);

var version$1 = "6.12.0";

const uploadMultipartWithProgress = (file, presigned, opts)=>Effect__namespace.gen(function*() {
        let uploadedBytes = 0;
        const etags = yield* Effect__namespace.forEach(presigned.urls, (url, index)=>{
            const offset = presigned.chunkSize * index;
            const end = Math.min(offset + presigned.chunkSize, file.size);
            const chunk = file.slice(offset, end);
            return uploadPart({
                url,
                chunk: chunk,
                contentDisposition: presigned.contentDisposition,
                fileType: file.type,
                fileName: file.name,
                onProgress: (delta)=>{
                    uploadedBytes += delta;
                    const percent = uploadedBytes / file.size * 100;
                    opts.onUploadProgress?.({
                        file: file.name,
                        progress: percent
                    });
                }
            }).pipe(Effect__namespace.andThen((tag)=>({
                    tag,
                    partNumber: index + 1
                })), Effect__namespace.retry({
                while: (error)=>error instanceof shared.RetryError,
                times: stdEnv.isTest ? 3 : 10,
                schedule: shared.exponentialBackoff()
            }));
        }, {
            concurrency: "inherit"
        }).pipe(Effect__namespace.tapErrorCause((error)=>opts.reportEventToUT("failure", {
                fileKey: presigned.key,
                uploadId: presigned.uploadId,
                fileName: file.name,
                storageProviderError: Cause__namespace.pretty(error).toString()
            })));
        // Tell the server that the upload is complete
        yield* opts.reportEventToUT("multipart-complete", {
            uploadId: presigned.uploadId,
            fileKey: presigned.key,
            etags
        });
    });
const uploadPart = (opts)=>Effect__namespace.async((resume)=>{
        const xhr = new XMLHttpRequest();
        xhr.open("PUT", opts.url, true);
        xhr.setRequestHeader("Content-Type", opts.fileType);
        xhr.setRequestHeader("Content-Disposition", shared.contentDisposition(opts.contentDisposition, opts.fileName));
        xhr.addEventListener("load", ()=>{
            const etag = xhr.getResponseHeader("Etag");
            if (xhr.status >= 200 && xhr.status <= 299 && etag) {
                return resume(Effect__namespace.succeed(etag));
            }
            return resume(Effect__namespace.fail(new shared.RetryError()));
        });
        xhr.addEventListener("error", ()=>resume(Effect__namespace.fail(new shared.RetryError())));
        let lastProgress = 0;
        xhr.upload.addEventListener("progress", (e)=>{
            const delta = e.loaded - lastProgress;
            lastProgress += delta;
            opts.onProgress(delta);
        });
        xhr.send(opts.chunk);
    });

const uploadPresignedPostWithProgress = (file, presigned, opts)=>Effect__namespace.async((resume)=>{
        const xhr = new XMLHttpRequest();
        xhr.open("POST", presigned.url);
        xhr.setRequestHeader("Accept", "application/xml");
        xhr.upload.addEventListener("progress", ({ loaded, total })=>{
            opts.onUploadProgress?.({
                file: file.name,
                progress: loaded / total * 100
            });
        });
        xhr.addEventListener("load", ()=>resume(xhr.status >= 200 && xhr.status < 300 ? Effect__namespace.succeed(null) : opts.reportEventToUT("failure", {
                fileKey: presigned.key,
                uploadId: null,
                fileName: file.name,
                storageProviderError: xhr.responseText
            })));
        xhr.addEventListener("error", ()=>resume(opts.reportEventToUT("failure", {
                fileKey: presigned.key,
                uploadId: null,
                fileName: file.name
            })));
        const formData = new FormData();
        Object.entries(presigned.fields).forEach(([k, v])=>formData.append(k, v));
        formData.append("file", file); // File data **MUST GO LAST**
        xhr.send(formData);
    });

const maybeParseResponseXML = (maybeXml)=>{
    const codeMatch = maybeXml.match(/<Code>(.*?)<\/Code>/s);
    const messageMatch = maybeXml.match(/<Message>(.*?)<\/Message>/s);
    const code = codeMatch?.[1];
    const message = messageMatch?.[1];
    if (!code || !message) return null;
    return {
        code: s3CodeToUploadThingCode[code] ?? DEFAULT_ERROR_CODE,
        message
    };
};
/**
 * Map S3 error codes to UploadThing error codes
 *
 * This is a subset of the S3 error codes, based on what seemed most likely to
 * occur in uploadthing. For a full list of S3 error codes, see:
 * https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html
 */ const DEFAULT_ERROR_CODE = "UPLOAD_FAILED";
const s3CodeToUploadThingCode = {
    AccessDenied: "FORBIDDEN",
    EntityTooSmall: "TOO_SMALL",
    EntityTooLarge: "TOO_LARGE",
    ExpiredToken: "FORBIDDEN",
    IncorrectNumberOfFilesInPostRequest: "TOO_MANY_FILES",
    InternalError: "INTERNAL_SERVER_ERROR",
    KeyTooLongError: "KEY_TOO_LONG",
    MaxMessageLengthExceeded: "TOO_LARGE"
};

const createAPIRequestUrl = (config)=>{
    const url = new URL(config.url);
    const queryParams = new URLSearchParams(url.search);
    queryParams.set("actionType", config.actionType);
    queryParams.set("slug", config.slug);
    url.search = queryParams.toString();
    return url;
};
/**
 * Creates a "client" for reporting events to the UploadThing server via the user's API endpoint.
 * Events are handled in "./handler.ts starting at L112"
 */ const createUTReporter = (cfg)=>(type, payload)=>Effect__namespace.gen(function*() {
            const url = createAPIRequestUrl({
                url: cfg.url,
                slug: cfg.endpoint,
                actionType: type
            });
            let headers = typeof cfg.headers === "function" ? cfg.headers() : cfg.headers;
            if (headers instanceof Promise) {
                headers = yield* Effect__namespace.promise(()=>headers);
            }
            const response = yield* shared.fetchEff(url, {
                method: "POST",
                body: JSON.stringify(payload),
                headers: {
                    "Content-Type": "application/json",
                    "x-uploadthing-package": cfg.package,
                    "x-uploadthing-version": version$1,
                    ...headers
                }
            }).pipe(Effect__namespace.andThen(shared.parseResponseJson), /**
         * We don't _need_ to validate the response here, just cast it for now.
         * As of now, @effect/schema includes quite a few bytes we cut out by this...
         * We have "strong typing" on the backend that ensures the shape should match.
         */ Effect__namespace.map(Function.unsafeCoerce), Effect__namespace.catchTags({
                FetchError: (e)=>new shared.UploadThingError({
                        code: "INTERNAL_CLIENT_ERROR",
                        message: `Failed to report event "${type}" to UploadThing server`,
                        cause: e
                    }),
                BadRequestError: (e)=>new shared.UploadThingError({
                        code: shared.getErrorTypeFromStatusCode(e.status),
                        message: e.getMessage(),
                        cause: e.json
                    }),
                InvalidJsonError: (e)=>new shared.UploadThingError({
                        code: "INTERNAL_CLIENT_ERROR",
                        message: "Failed to parse response from UploadThing server",
                        cause: e
                    })
            }));
            switch(type){
                case "failure":
                    {
                        // why isn't this narrowed automatically?
                        const p = payload;
                        const parsed = maybeParseResponseXML(p.storageProviderError ?? "");
                        if (parsed?.message) {
                            return yield* new shared.UploadThingError({
                                code: parsed.code,
                                message: parsed.message
                            });
                        } else {
                            return yield* new shared.UploadThingError({
                                code: "UPLOAD_FAILED",
                                message: `Failed to upload file ${p.fileName} to S3`,
                                cause: p.storageProviderError
                            });
                        }
                    }
            }
            return response;
        });

const version = version$1;
/**
 * Validate that a file is of a valid type given a route config
 * @public
 */ const isValidFileType = (file, routeConfig)=>Effect__namespace.runSync(shared.getTypeFromFileName(file.name, shared.objectKeys(routeConfig)).pipe(Effect__namespace.flatMap((type)=>Effect__namespace.succeed(file.type.includes(type))), Effect__namespace.catchAll(()=>Effect__namespace.succeed(false))));
/**
 * Validate that a file is of a valid size given a route config
 * @public
 */ const isValidFileSize = (file, routeConfig)=>Effect__namespace.runSync(shared.getTypeFromFileName(file.name, shared.objectKeys(routeConfig)).pipe(Effect__namespace.flatMap((type)=>shared.fileSizeToBytes(routeConfig[type].maxFileSize)), Effect__namespace.flatMap((maxFileSize)=>Effect__namespace.succeed(file.size <= maxFileSize)), Effect__namespace.catchAll(()=>Effect__namespace.succeed(false))));
const uploadFilesInternal = (endpoint, opts)=>{
    // classic service right here
    const reportEventToUT = createUTReporter({
        endpoint: String(endpoint),
        package: opts.package,
        url: shared.resolveMaybeUrlArg(opts.url),
        headers: opts.headers
    });
    return Effect__namespace.flatMap(reportEventToUT("upload", {
        input: "input" in opts ? opts.input : null,
        files: opts.files.map((f)=>({
                name: f.name,
                size: f.size,
                type: f.type
            }))
    }), Effect__namespace.forEach((presigned)=>uploadFile(String(endpoint), {
            ...opts,
            reportEventToUT
        }, presigned), {
        concurrency: 6
    }));
};
const genUploader = (initOpts)=>{
    return (endpoint, opts)=>uploadFilesInternal(endpoint, {
            ...opts,
            url: shared.resolveMaybeUrlArg(initOpts?.url),
            package: initOpts.package,
            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
            input: opts.input
        }).pipe(Effect__namespace.provideService(shared.FetchContext, {
            fetch: globalThis.fetch.bind(globalThis),
            baseHeaders: {
                "x-uploadthing-version": version$1,
                "x-uploadthing-api-key": undefined,
                "x-uploadthing-fe-package": initOpts.package,
                "x-uploadthing-be-adapter": undefined
            }
        }), Effect__namespace.runPromise).catch((error)=>{
            if (!Runtime__namespace.isFiberFailure(error)) throw error;
            throw Cause__namespace.squash(error[Runtime__namespace.FiberFailureCauseId]);
        });
};
const isPollingResponse = (input)=>{
    if (!shared.isObject(input)) return false;
    if (input.status === "done") return "callbackData" in input;
    return input.status === "still waiting";
};
const isPollingDone = (input)=>{
    return input.status === "done";
};
const uploadFile = (slug, opts, presigned)=>Arr__namespace.findFirst(opts.files, (file)=>file.name === presigned.fileName).pipe(Effect__namespace.tapError(()=>Console__namespace.error("No file found for presigned URL", presigned)), Effect__namespace.mapError(()=>new shared.UploadThingError({
            code: "NOT_FOUND",
            message: "No file found for presigned URL",
            cause: `Expected file with name ${presigned.fileName} but got '${opts.files.join(",")}'`
        })), Effect__namespace.tap((file)=>opts.onUploadBegin?.({
            file: file.name
        })), Effect__namespace.tap((file)=>"urls" in presigned ? uploadMultipartWithProgress(file, presigned, opts) : uploadPresignedPostWithProgress(file, presigned, opts)), Effect__namespace.zip(shared.fetchEff(presigned.pollingUrl, {
        headers: {
            authorization: presigned.pollingJwt
        }
    }).pipe(Effect__namespace.flatMap(shared.parseResponseJson), Effect__namespace.catchTag("BadRequestError", (e)=>Effect__namespace.fail(new shared.UploadThingError({
            code: shared.getErrorTypeFromStatusCode(e.status),
            message: e.message,
            cause: e
        }))), Effect__namespace.filterOrDieMessage(isPollingResponse, "received a non PollingResponse from the polling endpoint"), Effect__namespace.filterOrFail(isPollingDone, ()=>new shared.RetryError()), Effect__namespace.map(({ callbackData })=>callbackData), Effect__namespace.retry({
        while: (res)=>res instanceof shared.RetryError,
        schedule: shared.exponentialBackoff()
    }), Effect__namespace.when(()=>!opts.skipPolling), Effect__namespace.map(Option__namespace.getOrNull), Effect__namespace.map(Function.unsafeCoerce))), Effect__namespace.map(([file, serverData])=>({
            name: file.name,
            size: file.size,
            key: presigned.key,
            serverData,
            url: "https://utfs.io/f/" + presigned.key,
            customId: presigned.customId,
            type: file.type
        })));

Object.defineProperty(exports, "generateClientDropzoneAccept", {
  enumerable: true,
  get: function () { return shared.generateClientDropzoneAccept; }
});
Object.defineProperty(exports, "generateMimeTypes", {
  enumerable: true,
  get: function () { return shared.generateMimeTypes; }
});
Object.defineProperty(exports, "generatePermittedFileTypes", {
  enumerable: true,
  get: function () { return shared.generatePermittedFileTypes; }
});
exports.genUploader = genUploader;
exports.isValidFileSize = isValidFileSize;
exports.isValidFileType = isValidFileType;
exports.version = version;
